contracts/proxy/transparent/TransparentUpgradeableProxy.sol

```diff
@@ -106,11 +106,11 @@ contract TransparentUpgradeableProxy is ERC1967Proxy {
     /**
      * @dev Upgrade the implementation of the proxy.
      */
-    function _dispatchUpgradeTo() private returns (bytes memory) {
-        _requireZeroValue();
+    function _dispatchUpgradeToAndCall() private returns (bytes memory) {
+        bool forceCall = msg.value > 0;
```

Comment:
At that point, I would remove forceCall altogether. That would slightly reduce the size of the contract (and so the deployment cost)

I don't think anyone would ever do `upgradeToAndCall(newImplementation, "")` with some value, and expect the value to be passed to the receive function to for migrating.

If you are really worried about value being stuck:
- the contract is upgradeable, so there is always a path to recovery
- I'd be ok to doing `if (data.length == 0 && msg.value > 0) revert StupidUpdate();`

Reply:
> I don't think anyone would ever do `upgradeToAndCall(newImplementation, "")` with some value, and expect the value to be passed to the receive function to for migrating.

Why not? People do crazy stuff, there is nothing necessarily stupid about this.

Anyway, I added the revert that you suggested and set `forceCall = false`, though I'm not sure if that is what you meant by "remove forceCall altogether".

Reply:
You are saying that using the receive function with value is a valid (re)initialisation path ?

I know understand where all our misunderstandings come from. We are not targeting the same users.

Reply:
> though I'm not sure if that is what you meant by "remove forceCall altogether".

Right, I was thinking of removing it from the library.